### PR TITLE
ZK-4378: Provide a Checkbox new mold: Tristate

### PR DESCRIPTION
**Pull request by plijyfes**

### CODE REVIEW

File: zul/src/archive/web/js/zul/wgt/Checkbox.ts [hunk 1]
```diff
@@ -204,10 +203,24 @@ zul.wgt.Checkbox = zk.$extends(zul.LabelImageWidget, {
...
+	/** Returns the current state according to isIndeterminate() and isChecked().
+	 *
+	 * @return boolean or null
+	 * @since 9.0.0
+	 */
+	getState(): boolean | null {
+		if (this.isIndeterminate())
+			return null;
+		else
+			return this.isChecked() ? true : false;
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/wgt/Checkbox.ts]: return this.isChecked()

File: zul/src/org/zkoss/zul/Checkbox.java [hunk 2]
```diff
@@ -78,6 +79,9 @@ public boolean isIndeterminate() {
 	 * @since 8.6.0
 	 */
 	public void setIndeterminate(boolean indeterminate) {
+		String mold = getMold();
+		if ("switch".equals(mold) || "toggle".equals(mold))
+			throw new UiException("Checkbox switch/toggle mold not support indeterminate yet." + this);
```
Comment (scribetw) [inline zul/src/org/zkoss/zul/Checkbox.java]: does not

File: zul/src/org/zkoss/zul/Checkbox.java [hunk 3]
```diff
@@ -163,6 +167,18 @@ public void setChecked(boolean checked) {
 			}
 		}
 	}
+	
+	/** Returns the current state according to isIndeterminate() and isChecked().
+	 *
+	 * @return CHECKED, UNCHECKED or INDETERMINATE
+	 * @since 5.0.0
```
Comment (scribetw) [inline zul/src/org/zkoss/zul/Checkbox.java]: 9.0.0

File: zul/src/org/zkoss/zul/Checkbox.java [hunk 4]
```diff
@@ -223,6 +239,17 @@ protected boolean isChildable() {
 		return false;
 	}
 
+	public void setMold(String mold) {
+		if (this._indeterminate && ("switch".equals(mold) || "toggle".equals(mold)))
+			throw new UiException("Checkbox switch/toggle mold not support indeterminate yet." + this);
+		if (mold.equals("tristate") && _init) {
```
Comment (scribetw) [inline zul/src/org/zkoss/zul/Checkbox.java]: "tristate".equals

File: zul/src/org/zkoss/zul/Checkbox.java [hunk 5]
```diff
@@ -223,6 +239,17 @@ protected boolean isChildable() {
...
+			super.setMold(mold);
+			setIndeterminate(true);
```
Comment (scribetw) [inline zul/src/org/zkoss/zul/Checkbox.java]: need it?

File: zul/src/org/zkoss/zul/Checkbox.java [hunk 6]
```diff
@@ -260,8 +287,22 @@ public void service(org.zkoss.zk.au.AuRequest request, boolean everError) {
...
+				if (state == null) {
+					_indeterminate = true;
+					_checked = false;
+				} else if (state) {
+					_indeterminate = false;
+					_checked = true;
+				} else {
+					_indeterminate = false;
+					_checked = false;
+				}
```
Comment (scribetw) [inline zul/src/org/zkoss/zul/Checkbox.java]: _indeterminate = state == null;
_checked = state != null && state;

File: zktest/test/java/org/zkoss/zktest/zats/test2/F90_ZK_4378Test.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.openqa.selenium.Keys;
+import org.zkoss.zktest.zats.WebDriverTestCase;
+import org.zkoss.zktest.zats.ztl.Element;
+import org.zkoss.zktest.zats.ztl.JQuery;
+
+public class F90_ZK_4378Test extends WebDriverTestCase {
+	private final static String CHECKED = "CHECKED";
```
Comment (scribetw) [inline zktest/test/java/org/zkoss/zktest/zats/test2/F90_ZK_4378Test.java]: Checkbox.State.CHECKED.toString()

File: zul/src/archive/web/js/zul/wgt/Checkbox.ts [hunk 8]
```diff
@@ -204,10 +203,24 @@ zul.wgt.Checkbox = zk.$extends(zul.LabelImageWidget, {
...
+					this._clearStateClassName();
+					jq(this.$n()).addClass(this._getClassNameByState());
+				}
 			}
 		}
 	},
-
+	/** Returns the current state according to isIndeterminate() and isChecked().
+	 *
+	 * @return boolean or null
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/wgt/Checkbox.ts]: Boolean

File: zul/src/org/zkoss/zul/Checkbox.java [hunk 9]
```diff
@@ -223,6 +239,12 @@ protected boolean isChildable() {
 		return false;
 	}
 
+	public void setMold(String mold) {
+		if (this._indeterminate && ("switch".equals(mold) || "toggle".equals(mold)))
+			throw new UiException("Checkbox switch/toggle mold not support indeterminate yet." + this);
```
Comment (scribetw) [inline zul/src/org/zkoss/zul/Checkbox.java]: does

File: zul/src/org/zkoss/zul/Checkbox.java [hunk 10]
```diff
@@ -260,8 +282,14 @@ public void service(org.zkoss.zk.au.AuRequest request, boolean everError) {
 		final String cmd = request.getCommand();
 		if (cmd.equals(Events.ON_CHECK)) {
 			CheckEvent evt = CheckEvent.getCheckEvent(request);
-			_checked = evt.isChecked();
-			_indeterminate = false;
+			if ("tristate".equals(getMold())) {
+				Boolean state = evt.getState();
+				_indeterminate = state == null;
+				_checked = (state != null) && state;
```
Comment (scribetw) [inline zul/src/org/zkoss/zul/Checkbox.java]: order:
_c
_i

**Comment by scribetw** [comment]
@zkadmin rebase

**Comment by zkadmin** [comment]
I just pushed up the changes and merged, enjoy!

**Pull request merged**